packages/sources/layer2-sequencer-health/src/starkware.ts

```diff
@@ -0,0 +1,135 @@
+      {
+        contractAddress: config.starkwareConfig.dummyAccountAddress,
+        entrypoint: 'initialize',
+        calldata: [starkKeyPub, '0'],
+      },
+      undefined,
+      { maxFee: '0' },
+    ),
+    error: `Transaction receipt not received in ${config.timeoutLimit} milliseconds`,
+  })
+}
+
+/**
+ * The centralized Starkware sequencer is made up of two internal components.  The first is the gateway which is used to query the Sequencer state
+ * and the batcher which adds transactions to the pending block and processes the pending block to add it into the L2 chain.  The
+ * Sequencer is considered unhealthy if either of the components are down.
+ *
+ *  - The gateway is considered to be down if request to fetch pending block times out or if we get a 404
+ *  - The batcher is considered to be down if we a new block has not been produced AND no new transactions have been added to the
+ *  pending block within a given time interval.
+ * status when fetching
+ * @param networks
+ * @returns true if Sequencer is healthy
+ */
+export const checkStarkwareSequencerPendingTransactions = (): ((
+  config: ExtendedConfig,
+) => Promise<boolean>) => {
+  let recordedStarkwareState: StarwareState = {
+    lastUpdated: 0,
+    isSequencerHealthy: true,
+    lastBlockResponse: null,
+  }
+  return async (config: ExtendedConfig): Promise<boolean> => {
+    const delta = config.delta
+    const currentTime = Date.now()
+    if (
+      recordedStarkwareState.lastUpdated > 0 &&
+      currentTime - recordedStarkwareState.lastUpdated < delta
+    ) {
+      Logger.debug(
+        `Skipping check for Starkware Sequencer health as it has been less than ${delta} seconds since last check`,
+      )
+      return recordedStarkwareState.isSequencerHealthy
+    }
+    const { pendingBlockResponse } = await getPendingBlockFromGateway(config)
+    if (!pendingBlockResponse) {
+      recordedStarkwareState.isSequencerHealthy = false
+      recordedStarkwareState.lastUpdated = currentTime
+      return false
+    }
+
+    const isBatcherHealthy = checkBatcherHealthy(
+      recordedStarkwareState.lastBlockResponse,
+      pendingBlockResponse,
+    )
+
+    recordedStarkwareState = {
+      lastBlockResponse: pendingBlockResponse,
+      lastUpdated: currentTime,
+      isSequencerHealthy: isBatcherHealthy,
+    }
+    return isBatcherHealthy
+  }
+}
+
+const getPendingBlockFromGateway = async (
+  config: ExtendedConfig,
+): Promise<{
+  pendingBlockResponse: GetBlockResponse | null
+}> => {
+  let pendingBlockResponse = null
+  try {
+    pendingBlockResponse = await retry<GetBlockResponse>({
+      promise: async () => config.starkwareConfig.provider.getBlock('pending'),
+      retryConfig: config.retryConfig,
+    })
+  } catch (e: any) {
+    if (e.providerStatusCode === 504) {
+      Logger.warn(
+        `Request to fetch pending block timed out.  Status Code: ${e.providerStatusCode}.  Sequencer: UNHEALTHY`,
+      )
+    } else {
+      throw e
```

Comment:
Where is this error actually caught?

Even if it is, it's very hard to follow the code. I couldn't find who handles this. I suggest we should handled it here, and not throw.

Reply:
The error is caught [here](https://github.com/smartcontractkit/external-adapters-js/blob/develop/packages/sources/layer2-sequencer-health/src/endpoint/health.ts#L123).  Similar to this [comment](https://github.com/smartcontractkit/external-adapters-js/pull/2039#discussion_r942593110), this has been done this way as it was the general flow the EA took to check for the Sequencer health.  I do agree that this EA needs a general refactor but I'm worried that we're introducing too many changes in this PR.